### PR TITLE
[ROOT-10302] Move glew to ROOT builtins

### PR DESCRIPTION
**Pull request by oshadura**

### CODE REVIEW

**Comment by osschar** [comment]
Hey,

Why don't you fix the TGLIncludes.h as you need to? Then you don't have to change 100 files with their includes.

Cheers,
Matevz

**Comment by oshadura** [comment]
After moving glew to builtins, I realized that adding proper dependencies in root/graf3d/  directory (a new target GLEW) introduces a circular dependency inside root/graf3d/ between FTGL and RGL (circular dependency was reported by CMake). Initially using TGLIncludes.h was not correct, because since all headers were in the same place, dependencies were never tracked. Now with current approach TGLIncludes.h is not needed anymore.

**Comment by osschar** [comment]
Well, it was never needed as such ... it was just to make includes simpler and avoid ifdefs in every TGL file :) I'm just saying you had to change every file that used it and this is really not necessary.

File: graf2d/cocoa/src/TGCocoa.mm [hunk 1]
```diff
@@ -11,13 +11,22 @@
 
 //#define NDEBUG
 
-#include "TGCocoa.h"
+#ifdef WIN32
```
Comment (Axel-Naumann) [inline graf2d/cocoa/src/TGCocoa.mm]: I don't understand what Windows has to do with Cocoa?

File: graf3d/CMakeLists.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.
+  # Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.
```
Comment (Axel-Naumann) [inline graf3d/CMakeLists.txt]: Why the extra space?

File: graf3d/eve/CMakeLists.txt [hunk 3]
```diff
@@ -228,7 +228,9 @@ if(MSVC)
       OpenGL::GL
       OpenGL::GLU
       ${FTGL_LIBRARIES}
-      ${GLEW_LIBRARIES}
+      GLEW
```
Comment (Axel-Naumann) [inline graf3d/eve/CMakeLists.txt]: Why do we need to repeat `GLEW`? This looks like a deficiency in `ROOT_STANDARD_PACKAGE` checking whether a dependency is a builtin?

File: graf2d/cocoa/src/TGCocoa.mm (hunk 1 above)
Comment (oshadura) [reply graf2d/cocoa/src/TGCocoa.mm]: I was talking with Bertand, I will check if it is still needed (it was a part of TGLInclude.h file)

File: graf3d/CMakeLists.txt (hunk 2 above)
Comment (oshadura) [reply graf3d/CMakeLists.txt]: Thanks!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@phsft-bot build!

**Comment by oshadura** [comment]
@Axel-Naumann I still would like to land it. Can please let me know what is a blocker here?

**Comment by oshadura** [comment]
@osschar I would like to know your opinion about this PR? If you want we can have a chat over Skype...One of benefits, is that we will close Jira issue about an upgrade of glew version and it helps to remove the circular dependency.

**Comment by osschar** [comment]
it's the same thing i said earlier ... you have a cut-n-paste of
```
 #ifdef WIN32
#include "Windows4Root.h"
#endif

#include <GL/glew.h>
```
in close to hundred files ... and you removed usage of TGLIncludes.h that had this (or similar) before.

If you put this back into TGLIncludes, all graf3d/gl files will remain unchanged.

**Comment by oshadura** [comment]
@osschar the problem that it will not work...

Having TGLIncludes.h (with #include <GL/glew.h>) make RGL depends on GLEW, and then as a consequence, I need to put graf3d\ftgl (FTGL) depends on RGL. Having this is creating circular dependency:

`CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "FTGL" of type SHARED_LIBRARY
    depends on "RGL" (weak)
  "RGL" of type SHARED_LIBRARY
    depends on "FTGL" (weak)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
CMake Generate step failed.  Build files cannot be regenerated correctly.
make: *** [Makefile:9096: cmake_check_build_system] Error 1`

Why it was not visible before because TGLIncludes.h were copied in ${CMAKE_BINARY_DIR}/include/ where it is was picked up by whatever ROOT library you were building before.  To achieve better modularity is could be fixed, in a way, I am proposing in this PR.

-------------------------------------------------------------------------------------
Here is a test PR I opened  to show you an error for your proposal:  https://github.com/root-project/root/pull/4786

`C:\build\workspace\root-pullrequests-build\build\include\TGLIncludes.h(21,10): fatal error C1083: Cannot open include file: 'GL/glew.h': No such file or directory [C:\build\workspace\root-pullrequests-build\build\graf3d\ftgl\FTGL.vcxproj]`

**Comment by oshadura** [comment]
What I can try is to remove 
`
#ifdef WIN32
#include "Windows4Root.h"
#endif
`
Maybe this code is redundant?

**Comment by oshadura** [comment]
@phsft-bot build just on windows10/cxx14

**Comment by oshadura** [comment]
> What I can try is to remove
> 
> ```
> #ifdef WIN32
> #include "Windows4Root.h"
> #endif
> ```
> 
> Maybe this code is redundant?

@Axel-Naumann @osschar 
I tried and actually it used/needed widely over graf3d, so I will leave it like this! Let me know if you have any ideas how to improve it, otherwise I think it is ready to go.

Please note this PR resolves as well https://sft.its.cern.ch/jira/browse/ROOT-10302 issue

**Comment by osschar** [comment]
> @osschar the problem that it will not work...
> 
> Having TGLIncludes.h (with #include <GL/glew.h>) make RGL depends on GLEW, and then as a consequence, I need to put graf3d\ftgl (FTGL) depends on RGL. Having this is creating circular dependency:
> 
> ```
> CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
>   "FTGL" of type SHARED_LIBRARY
>     depends on "RGL" (weak)
>   "RGL" of type SHARED_LIBRARY
>     depends on "FTGL" (weak)
> At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
> CMake Generate step failed.  Build files cannot be regenerated correctly.
> make: *** [Makefile:9096: cmake_check_build_system] Error 1
> ```
> 
> Why it was not visible before because TGLIncludes.h were copied in ${CMAKE_BINARY_DIR}/include/ where it was picked up by whatever ROOT library you were building before. To achieve better modularity this could be fixed, in a way, I am proposing in this PR.
> 
> Here is a test PR I opened to show you an error for your proposal: #4786
> 
> ```
> C:\build\workspace\root-pullrequests-build\build\include\TGLIncludes.h(21,10): fatal error C1083: Cannot open include file: 'GL/glew.h': No such file or directory [C:\build\workspace\root-pullrequests-build\build\graf3d\ftgl\FTGL.vcxproj]
> ```

Oh, but why don't you just make the change to direct GL/glew.h inclusion in FTGL then?

**Comment by osschar** [comment]
> What I can try is to remove
> 
> ```
> #ifdef WIN32
> #include "Windows4Root.h"
> #endif
> ```
> 
> Maybe this code is redundant?

I'd have guessed it is needed (as we now know :) ) ... the windows part was maintained by Timur so I don't know what's real issue behind and if there was any change for windows 10. I do know however that there is pretty big difference in WGL and GLX implementation for RGL rendering contexts and render surfaces. Probably something for Cocoa as well, another thing that was done by Timur :(

**Comment by oshadura** [comment]
> > @osschar the problem that it will not work...
> > Having TGLIncludes.h (with #include <GL/glew.h>) make RGL depends on GLEW, and then as a consequence, I need to put graf3d\ftgl (FTGL) depends on RGL. Having this is creating circular dependency:
> > ```
> > CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
> >   "FTGL" of type SHARED_LIBRARY
> >     depends on "RGL" (weak)
> >   "RGL" of type SHARED_LIBRARY
> >     depends on "FTGL" (weak)
> > At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
> > CMake Generate step failed.  Build files cannot be regenerated correctly.
> > make: *** [Makefile:9096: cmake_check_build_system] Error 1
> > ```
> > 
> > 
> > Why it was not visible before because TGLIncludes.h were copied in ${CMAKE_BINARY_DIR}/include/ where it was picked up by whatever ROOT library you were building before. To achieve better modularity this could be fixed, in a way, I am proposing in this PR.
> > Here is a test PR I opened to show you an error for your proposal: #4786
> > ```
> > C:\build\workspace\root-pullrequests-build\build\include\TGLIncludes.h(21,10): fatal error C1083: Cannot open include file: 'GL/glew.h': No such file or directory [C:\build\workspace\root-pullrequests-build\build\graf3d\ftgl\FTGL.vcxproj]
> > ```
> 
> Oh, but why don't you just make the change to direct GL/glew.h inclusion in FTGL then?

Because it is indirect dependency :( If you check files that I changed, there is no direct includes of glew.h

**Comment by oshadura** [comment]
@linev can you please validate with files doesn't need (I removed from TGCocoa.mm) this:
`#ifdef WIN32
#include "Windows4Root.h"
#endif`

Can we try find some agreement and move forward with this PR?

**Comment by linev** [comment]
Logically one do not require windows includes in cocoa. But probably @bellenot can give more comments as Windows expert

**Comment by bellenot** [comment]
> Logically one do not require windows includes in cocoa. But probably @bellenot can give more comments as Windows expert

Indeed, cocoa and Windows are two different (and unrelated) platforms...

**Pull request closed without merging**